File: python/1d-dynamic-programming/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        array = [1] * len(nums)
        for i in range(len(nums) - 2, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j]:
                    array[i] = max(array[j] + 1, array[i])

        return max(array)

File: c++/1d-dynamic-programming/longest-increasing-subsequence.cc
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        std::vector<int> lis(nums.size(), 1);
        for (int j = nums.size() - 1; j > -1; --j) {
            for (int i = j + 1; i < nums.size(); ++i) {
                if (nums[j] < nums[i]) {
                    lis[j] = std::max(lis[j], lis[i] + 1);
                }
            }
        }

        int answer = 1;
        for (int result : lis) {
            answer = std::max(answer, result);
        }

        return answer;
    }
};
